Eastern-Research-Group/expert-query
# Feature/89 170 performance improvments

I was thinking about adding more performance improvements to this, but I decided that this was a good enough starting point. I want to see how long it takes to generate all of these indexes.

## Related Issues:
* [EQ-89](https://jira.epa.gov/browse/EQ-89)
* [EQ-170](https://jira.epa.gov/browse/EQ-170)

## Main Changes:
* Added indexes to any columns we plan to query against.
* Switched from using distinct queries to using recursive queries. This improves performance by emulating a loose scan index (currently not supported by postgres).
* Removed the `assessmentUnitStatus` field from tmdl, because this field is not in our tmdl table.
* Updated how our counts work. Apparently it is faster to do `count(*)` than it is to do `count(primary_key_column)`. 

## Steps To Test:
1. Add `CREATE_INDEXES_AFTER_DATA=true` to your `etl/.env.local` file
2. Run `npm run etl_database`
3. Verify the indexes were added
4. Verify the app runs

## TODO:
* [ ] Test the index creation in cloud.gov
* [ ] Get some index creation timing information from the logs
* [ ] Do some testing to see if there are indexes we can delete
* [ ] Further tune performance of commonly run queries (could involve creating materialized views)
